### PR TITLE
Backport functions required by Comment Query Loop and related blocks

### PR DESCRIPTION
**Pull request by DAreRodz**
Backport changes from Gutenberg to add functions required by Comment Query Loop and related blocks.

From [`/lib/compat/wordpress-6.0/blocks.php`](https://github.com/WordPress/gutenberg/blob/6f7c5b573a6023c2f2ce3eecd0d6b169ce4e964c/lib/compat/wordpress-6.0/blocks.php):

- [x] `build_comment_query_vars_from_block` function
- [x] `get_comments_pagination_arrow` function
- [x] `gutenberg_extend_block_editor_settings_with_discussion_settings` filter's callback
- [x] `gutenberg_rest_comment_set_children_as_embeddable` filter's callback

Tests:
- [ ] [`class-block-library-comment-template-test.php`](https://github.com/WordPress/gutenberg/blob/6a05ccc9a5b842b6e375ba6c1afbba8397abb7db/phpunit/class-block-library-comment-template-test.php) unit test

Also being tracked in https://github.com/WordPress/gutenberg/issues/39889.

Trac ticket: https://core.trac.wordpress.org/ticket/55505 and https://core.trac.wordpress.org/ticket/55567

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/block-library/commentTemplate.php [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+ */
+
+/**
+ * Tests for the Comment Template block.
+ *
+ * @since 6.0.0
+ *
+ * @group block-library
+ */
+class Test_Block_Library_CommentTemplate extends WP_UnitTestCase {
```
Comment (DAreRodz) [inline tests/phpunit/tests/block-library/commentTemplate.php:17]: I've defined a new `@subpackage` and `@group` for this test, as this is the first one we have for a specific block (as far as I saw).

I've also modified the class name to follow the same format as the other test files.

Comment (DAreRodz) [reply tests/phpunit/tests/block-library/commentTemplate.php:17]: Let me know if this file should be placed under a different folder/subpackage/etc. 🙂

**Comment by DAreRodz** [comment]
Uh, there's a missing function, I'm taking a look.

File: src/wp-includes/blocks.php [hunk 2]
```diff
@@ -1358,3 +1358,153 @@ function _wp_multiple_block_styles( $metadata ) {
...
+ *
+ * @since 6.0.0
+ *
+ * @param array $settings Default editor settings.
+ *
+ * @return array Filtered editor settings.
+ */
+function extend_block_editor_settings_with_discussion_settings( $settings ) {
+
+	$settings['__experimentalDiscussionSettings'] = array(
```
Comment (gziolo) [inline src/wp-includes/blocks.php]: This needs to go somewhere before this line:

https://github.com/WordPress/wordpress-develop/blob/b6394c2e9fcdea317ff350d59e985a40e02edd41/src/wp-includes/block-editor.php#L407

We don't need to use the hook in core anymore.

File: src/wp-includes/blocks.php [hunk 3]
```diff
@@ -1358,3 +1358,153 @@ function _wp_multiple_block_styles( $metadata ) {
...
+ *
+ * @since 6.0.0
+ *
+ * @return void
+ */
+function gutenberg_rest_comment_set_children_as_embeddable() {
+	add_filter(
+		'rest_prepare_comment',
+		function ( $response ) {
+			$links = $response->get_links();
```
Comment (gziolo) [inline src/wp-includes/blocks.php]: This part also should land in the class that handles the REST API endpoint for comments. @spacedmonkey and @TimothyBJacobs will know best where to move it.

Comment (TimothyBJacobs) [reply src/wp-includes/blocks.php]: Yeah, this needs to move into `WP_REST_Comments_Controller::prepare_links()`.

File: tests/phpunit/tests/block-library/commentTemplate.php [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+			1,
+			array(
+				'comment_author'       => 'Test',
+				'comment_author_email' => 'test@example.org',
+				'comment_content'      => 'Hello world',
+			)
+		);
+	}
+
+	function test_build_comment_query_vars_from_block_with_context() {
```
Comment (DAreRodz) [inline tests/phpunit/tests/block-library/commentTemplate.php:52]: I noticed this test is failing in my local environment. I'm taking a look into that. 👀 

```
1) Test_Block_Library_CommentTemplate::test_build_comment_query_vars_from_block_with_context
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
     'status' => 'approve'
     'no_found_rows' => false
     'update_comment_meta_cache' => false
-    'post_id' => 4
     'hierarchical' => 'threaded'
     'number' => 5
     'paged' => 1
 )
```

Comment (gziolo) [reply tests/phpunit/tests/block-library/commentTemplate.php:52]: I see the same error locally. It's really surprising.

**Comment by gziolo** [comment]
I committed all the changes but tests with https://core.trac.wordpress.org/changeset/53138 to unblock https://github.com/WordPress/wordpress-develop/pull/2564.

File: tests/phpunit/tests/block-library/commentTemplate.php (hunk 1 above)
Comment (gziolo) [reply tests/phpunit/tests/block-library/commentTemplate.php:17]: It's fine as is. It's an implementation detail I can look at later.

File: tests/phpunit/tests/block-library/commentTemplate.php (hunk 4 above)
Comment (DAreRodz) [reply tests/phpunit/tests/block-library/commentTemplate.php:52]: For some reason, the block is not being instantiated here with context; it's empty.

https://github.com/WordPress/wordpress-develop/blob/537d27225db483c6c16afc1daef6364d4f7668e4/tests/phpunit/tests/block-library/commentTemplate.php#L57-L62

I see the same happening in all the tests where context is used. Could it be related to the block not included in WordPress core yet? 🤔

Comment (gziolo) [reply tests/phpunit/tests/block-library/commentTemplate.php:52]: I think I know why, we need to have the block code first and we can then land the tests. All good then. Let's wait for https://github.com/WordPress/wordpress-develop/pull/2564 and see how it looks afterwards. Good detective work 😄

**Comment by gziolo** [comment]
I backported 3 out of 5 unit tests to WordPress core with https://core.trac.wordpress.org/changeset/53172.

I had some issues with:

- `test_rendering_comment_template`
- `test_rendering_comment_template_nested`

I would appreciate some help. It can be another PR if that makes it easier.

**Comment by DAreRodz** [comment]
Hey, @gziolo, sure! Let me create a new PR for that. I think it would be easier as there are some changes already integrated into `trunk`.

**Comment by DAreRodz** [comment]
> I had some issues with:
> 
> * `test_rendering_comment_template`
> * `test_rendering_comment_template_nested`

The problem was using

```php
render_block_core_comment_template( null, null, $block )
```

 instead of

```php
$block->render()
```

in the failing tests.

The second call returns (almost) the same output, but also updates ―among other things― `WP_Block_Supports::$block_to_render`, which should be not `null` when `get_block_wrapper_attributes()` is called. So, I guess `render_block_core_comment_template()` is not meant to be called directly.

The first call works in Gutenberg, though; I don't know why. 🤷 

I've created a new PR here with the test fixes: https://github.com/WordPress/wordpress-develop/pull/2613. I can close this one if you feel so. ☺️

cc: @gziolo

**Comment by gziolo** [comment]
> I've created a new PR here with the test fixes: https://github.com/WordPress/wordpress-develop/pull/2613. I can close this one if you feel so. ☺️

Awesome, thank you so much 💯

File: src/wp-includes/blocks.php [hunk 5]
```diff
@@ -1358,3 +1358,96 @@ function _wp_multiple_block_styles( $metadata ) {
...
+			$comment_args['paged'] = $page;
+		} elseif ( 'oldest' === $default_page ) {
+			$comment_args['paged'] = 1;
+		} elseif ( 'newest' === $default_page ) {
+			$comment_args['paged'] = (int) ( new WP_Comment_Query( $comment_args ) )->max_num_pages;
+		}
+		// Set the `cpage` query var to ensure the previous and next pagination links are correct
+		// when inheriting the Discussion Settings.
+		if ( 0 === $page && isset( $comment_args['paged'] ) && $comment_args['paged'] > 0 ) {
+			set_query_var( 'cpage', $comment_args['paged'] );
```
Comment (sybrew) [inline src/wp-includes/blocks.php:1412]: Why write to a query variable here? It breaks everything else in WP. It should remain `0` on the first page, as it always had been.

Comment (ockham) [reply src/wp-includes/blocks.php:1412]: Cross-referencing https://github.com/WordPress/wordpress-develop/pull/6291, where this is now being discussed, and https://github.com/WordPress/gutenberg/pull/39227, where it was originally introduced.

**Pull request closed without merging**